### PR TITLE
Add WCP changes to AKO master

### PR DESCRIPTION
**Pull request by abhishekbsingh**

### CODE REVIEW

**Comment by abhishekbsingh** [comment]
build ako

**Comment by abhishekbsingh** [comment]
build ako

**Comment by abhishekbsingh** [comment]
build ako

**Comment by abhishekbsingh** [comment]
build ako

**Comment by abhishekbsingh** [comment]
build ako

**Comment by abhishekbsingh** [comment]
build ako

**Comment by abhishekbsingh** [comment]
build ako

**Comment by abhishekbsingh** [comment]
build ako

**Comment by abhishekbsingh** [comment]
build ako

**Comment by abhishekbsingh** [comment]
build ako

**Comment by abhishekbsingh** [comment]
build ako

File: cmd/ako-main/main.go [hunk 1]
```diff
@@ -172,11 +172,24 @@ func InitializeAKC() {
 	}
 
 	akoControlConfig.SetEventRecorder(lib.AKOEventComponent, kubeClient, false)
-	pod, err := kubeClient.CoreV1().Pods(utils.GetAKONamespace()).Get(context.TODO(), os.Getenv("POD_NAME"), metav1.GetOptions{})
-	if err != nil {
-		utils.AviLog.Warnf("Error getting AKO pod details, %s.", err.Error())
+
+	// POD_NAME is not set in case of a WCP cluster
+	if os.Getenv("POD_NAME") == "" {
+		pods, err := kubeClient.CoreV1().Pods(utils.GetAKONamespace()).List(context.TODO(), metav1.ListOptions{Limit: 1})
```
Comment (akshayhavile) [inline cmd/ako-main/main.go:178]: what adv. do we get by setting up limit to 1?

File: cmd/ako-main/main.go [hunk 2]
```diff
@@ -172,11 +172,24 @@ func InitializeAKC() {
...
+		if err != nil {
+			utils.AviLog.Warnf("Error getting AKO pod details, %s.", err.Error())
+		} else {
+			for _, pod := range pods.Items {
+				akoControlConfig.SaveAKOPodObjectMeta(&pod)
```
Comment (akshayhavile) [inline cmd/ako-main/main.go:183]: Do we want to save info of all pods deployed in vmware system namespace?

File: ako-clean/cleanup.go [hunk 3]
```diff
@@ -346,6 +348,109 @@ func cleanupPools() error {
...
+}
+
+func sanitzeAviCloud() error {
+	aviClient := avicache.SharedAVIClients(lib.GetAdminTenant()).AviClient[0]
+	err := avirest.AviCloudCachePopulate(aviClient, utils.CloudName)
+	if err != nil {
+		return err
+	}
+	dataNetworkTier1Lrs := make([]*models.Tier1LogicalRouterInfo, 0)
+	cloudTier1Lrs := avirest.CloudCache.NsxtConfiguration.DataNetworkConfig.Tier1SegmentConfig.Manual.Tier1Lrs
```
[the remark this answers is not in the document]
Comment (abhishekbsingh) [reply ako-clean/cleanup.go:388]: They aren't expected to be null. This script is triggered as part of WCP disablement workflow where the cloud and other avi resources are expected to be present.

File: ako-clean/cleanup.go [hunk 4]
```diff
@@ -346,6 +348,109 @@ func cleanupPools() error {
...
+				continue
+			}
+			networkRef := "/api/network/?name=" + netName
+			usableNetworks = append(usableNetworks, &models.IPAMUsableNetwork{NwRef: &networkRef})
+		}
+	}
+	if !updateIPAM {
+		return nil
+	}
+	ipam.InternalProfile.UsableNetworks = usableNetworks
```
[the remark this answers is not in the document]
Comment (abhishekbsingh) [reply ako-clean/cleanup.go:441]: That is outside the scope of this script. It should only remove the Networks which are specific to a particular Supervisor cluster

File: ako-infra/avirest/netinfo_handler.go [hunk 5]
```diff
@@ -152,47 +152,39 @@ func (t *T1LRNetworking) createInfraSettingAndAnnotateNS(nsLRMap map[string]stri
...
+		_, err := lib.CreateOrUpdateAviInfraSetting(infraSettingName, netName, lr)
+		if err != nil {
+			utils.AviLog.Errorf("failed to create aviInfraSetting, name: %s, error: %s", infraSettingName, err.Error())
+			continue
+		}
+		lib.AnnotateNamespaceWithInfraSetting(ns, infraSettingName)
 	}
 
 	for infraSettingName := range staleInfraSettingCRSet {
-		wg.Add(1)
```
[the remark this answers is not in the document]
Comment (abhishekbsingh) [reply ako-infra/avirest/netinfo_handler.go:186]: Goroutines could lead to throttling from kubernetes. It is better to not use goroutines here.

File: ako-infra/avirest/vpc_handler.go [hunk 6]
```diff
@@ -49,76 +45,12 @@ func (v *VPCHandler) AddNetworkInfoEventHandler(stopCh <-chan struct{}) {
...
 
 func (v *VPCHandler) SyncLSLRNetwork() {
-	vpcToSubnetMap, nsToVPCMap, err := lib.GetVPCs()
+	nsToVPCMap, err := lib.GetVPCs()
 	if err != nil {
 		utils.AviLog.Errorf("Failed to list VPCs, error: %s", err)
 		return
 	}
 	utils.AviLog.Infof("Got NS to VPC Map: %v", nsToVPCMap)
-	client := InfraAviClientInstance()
```
[the remark this answers is not in the document]
Comment (abhishekbsingh) [reply ako-infra/avirest/vpc_handler.go:58]: Yes, Avi Controller has added support for streaming NSX events and will add VPC/Subnets based on those events. No need to push from AKO anymore.

File: ako-infra/ingestion/avi_infra_controller.go [hunk 7]
```diff
@@ -108,12 +115,17 @@ func (a *AviControllerInfra) DeriveCloudNameAndSEGroupTmpl(tz string) (error, st
 		if *cloud.Vtype != lib.CLOUD_NSXT || cloud.NsxtConfiguration == nil {
 			continue
 		}
-		if cloud.NsxtConfiguration.VpcMode != nil && *cloud.NsxtConfiguration.VpcMode != lib.GetVPCMode() {
+		if lib.GetVPCMode() && (cloud.NsxtConfiguration.VpcMode == nil || !*cloud.NsxtConfiguration.VpcMode) {
```
[the remark this answers is not in the document]
Comment (abhishekbsingh) [reply ako-infra/ingestion/avi_infra_controller.go:118]: I hadn't done that as I am assuming it shouldn't be very expensive to call GetVPCMode, but will add it to a variable and reuse that.

File: cmd/ako-main/main.go (hunk 1 above)
Comment (abhishekbsingh) [reply cmd/ako-main/main.go:178]: No advantage as such, just wanted to get 1 pod.

File: cmd/ako-main/main.go (hunk 2 above)
Comment (abhishekbsingh) [reply cmd/ako-main/main.go:183]: It would be a single pod due to Limit being set to 1

**Comment by abhishekbsingh** [comment]
build ako

**Comment by abhishekbsingh** [comment]
build ako

**Comment by abhishekbsingh** [comment]
build ako

**Review by akshayhavile — APPROVED** [review]
@abhishekbsingh LGTM. with one minor comment

**Comment by abhishekbsingh** [comment]
> Will this functionality available for different versions of Avi controller? like 22.1.x, 30.1.x, 31.1.x etc.??
This would be available on in 31.1.1 and onwards controller versions.

**Review by pkoshtavmware — APPROVED** [review]
LGTM with some minor comments

**Pull request merged**